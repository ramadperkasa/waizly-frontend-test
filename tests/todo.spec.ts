import { describe, it, expect, beforeEach } from "vitest";
import { mount } from "@vue/test-utils";

import TodoAddData from "../components/todo/add-data.vue";
import DataNotFound from "../components/todo/data-not-found.vue";
import Search from "../components/todo/search.vue";

import Index from "../pages/index.vue";

describe("TodoAddData", () => {
  it("should render the component", () => {
    const wrapper = mount(TodoAddData);
    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe("DataNotFound", () => {
  it("should render the component", () => {
    const wrapper = mount(DataNotFound);
    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe("Search", () => {
  it("should render the component", () => {
    const wrapper = mount(Search);
    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe("Index", () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = mount(Index);
  });

  it("is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("initial state is correct", () => {
    expect(wrapper.vm.todoList).toEqual([]);
    expect(wrapper.vm.todoInput).toBe("");
    expect(wrapper.vm.searchTodo).toBe("");
  });

  it("adds a new todo item correctly", async () => {
    await wrapper.setData({ todoInput: "New Todo" });
    await wrapper.vm.onSubmit();
    expect(wrapper.vm.todoList).toEqual([
      { id: 1, name: "New Todo", isDone: false, isEdit: false },
    ]);
  });

  it("removes a todo item correctly", async () => {
    await wrapper.setData({
      todoList: [{ id: 1, name: "New Todo", isDone: false, isEdit: false }],
    });
    await wrapper.vm.onRemove(0);
    expect(wrapper.vm.todoList).toEqual([]);
  });

  it("toggles isDone property of a todo item correctly", async () => {
    await wrapper.setData({
      todoList: [{ id: 1, name: "New Todo", isDone: false, isEdit: false }],
    });
    await wrapper.vm.onToggleCheck(0);
    expect(wrapper.vm.todoList[0].isDone).toBe(true);
  });

  it("filters the todo list correctly", async () => {
    await wrapper.setData({
      todoList: [
        { id: 1, name: "First Todo", isDone: false, isEdit: false },
        { id: 2, name: "Second Todo", isDone: false, isEdit: false },
      ],
      searchTodo: "First",
    });
    expect(wrapper.vm.filteredTodo).toEqual([
      { id: 1, name: "First Todo", isDone: false, isEdit: false },
    ]);
  });
});
