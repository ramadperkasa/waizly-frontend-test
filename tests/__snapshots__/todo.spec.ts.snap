// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DataNotFound > should render the component 1`] = `
"<div class="flex flex-col">
  <div class="mx-auto">No Data Found! Please add todo</div><img src="/assets/images/no-data.svg" class="w-1/2 mx-auto" alt="No data found">
</div>"
`;

exports[`Search > should render the component 1`] = `
"<div class="flex justify-between align-center">
  <div>TODO LIST</div>
  <div class="w-48"><label for="search" class="mb-2 text-sm font-medium text-gray-900 sr-only dark:text-white">Search</label>
    <div class="relative">
      <div class="absolute inset-y-0 flex items-center pointer-events-none start-0 ps-3"><svg class="w-4 h-4 text-gray-500 dark:text-gray-400" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 20 20">
          <path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="m19 19-4-4m0-7A7 7 0 1 1 1 8a7 7 0 0 1 14 0Z"></path>
        </svg></div><input type="search" id="search" class="block w-full text-xs text-gray-900 border border-gray-300 rounded-lg ps-10 bg-gray-50 focus:ring-primary focus:border-primary dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-primary dark:focus:border-primary" placeholder="Search">
    </div>
  </div>
</div>"
`;

exports[`TodoAddData > should render the component 1`] = `
"<div class="relative">
  <div class="absolute inset-y-0 flex items-center pointer-events-none start-0 ps-3"><svg class="w-4 h-4 text-gray-500 dark:text-gray-400" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512">
      <path d="M40 48C26.7 48 16 58.7 16 72v48c0 13.3 10.7 24 24 24H88c13.3 0 24-10.7 24-24V72c0-13.3-10.7-24-24-24H40zM192 64c-17.7 0-32 14.3-32 32s14.3 32 32 32H480c17.7 0 32-14.3 32-32s-14.3-32-32-32H192zm0 160c-17.7 0-32 14.3-32 32s14.3 32 32 32H480c17.7 0 32-14.3 32-32s-14.3-32-32-32H192zm0 160c-17.7 0-32 14.3-32 32s14.3 32 32 32H480c17.7 0 32-14.3 32-32s-14.3-32-32-32H192zM16 232v48c0 13.3 10.7 24 24 24H88c13.3 0 24-10.7 24-24V232c0-13.3-10.7-24-24-24H40c-13.3 0-24 10.7-24 24zM40 368c-13.3 0-24 10.7-24 24v48c0 13.3 10.7 24 24 24H88c13.3 0 24-10.7 24-24V392c0-13.3-10.7-24-24-24H40z"></path>
    </svg></div><input type="search" id="default-search" class="block w-full p-4 text-sm text-gray-900 border border-gray-300 rounded-lg ps-10 bg-gray-50 focus:ring-primary/40 focus:border-primary dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-primary/40 dark:focus:border-primary" placeholder="What do you need to do"><button class="text-white absolute end-2.5 bottom-2.5 bg-primary hover:bg-primary/90 focus:ring-4 focus:outline-none focus:ring-primary/30 font-medium rounded-lg text-sm px-4 py-2 dark:bg-primary/80 dark:hover:bg-primary dark:focus:ring-primary/90hover:bg-primary/90"> Add </button>
</div>"
`;
